Feature/invitation

CptBreeza opened the pull request.

fahchen committed:
Change schema_format to default

[Issue #1] https://github.com/iec-uestc/campo/issues/1

https://trello.com/c/vjLL1u9r

fahchen committed:
Create model and relation with user to invitation

[Issue #1] https://github.com/iec-uestc/campo/issues/1

https://trello.com/c/vjLL1u9r

Create a model for invitation and build the relations between users and
invitations

fahchen committed:
Add generate to invitation

[Issue #1] https://github.com/iec-uestc/campo/issues/1

    https://trello.com/c/vjLL1u9r

Add a method to generate a code before the invitation is created.

fahchen committed:
Improve Invitation.generate

[Issue #1] https://github.com/iec-uestc/campo/issues/1

        https://trello.com/c/vjLL1u9r

Improved Invitation.generate so now it will generate a unique code

CptBreeza committed:
Support user to generate invitation

[Issue #1] https://github.com/iec-uestc/campo/issues/1

https://trello.com/c/vjLL1u9r

CptBreeza committed:
Add relations to inviters and invitees

[Issue #1] https://github.com/iec-uestc/campo/issues/1

https://trello.com/c/vjLL1u9r

CptBreeza committed:
Add invitation support to users

[Issue #1] https://github.com/iec-uestc/campo/issues/1

https://trello.com/c/vjLL1u9r

Add methods to users to check,set and generate invitation

CptBreeza committed:
Invitation needed now for user's signup

[Issue #1] https://github.com/iec-uestc/campo/issues/1

https://trello.com/c/vjLL1u9r

User now need a invitation code to sign up

CptBreeza committed:
Add support for users to get invitations

[Issue #1] https://github.com/iec-uestc/campo/issues/1

https://trello.com/c/vjLL1u9r

Users now can get a random number of invitation codes and invite their
friends to join!

CptBreeza committed:
Add support to admin to generate invitation codes

[Issue #1] https://github.com/iec-uestc/campo/issues/1

https://trello.com/c/vjLL1u9r

Allow administrator to generate infinite invitation codes,and fix the
language file of users.

CptBreeza committed:
Fix language files

[Issue #1] https://github.com/iec-uestc/campo/issues/1

https://trello.com/c/vjLL1u9r

Fix language files of users and administrators.

CptBreeza committed:
Not update attributes when invitation not existing

[Issue #1] https://github.com/iec-uestc/tsemple/issues/1

https://trello.com/c/vjLL1u9r

Not update attributes when creating a user without invitation code

CptBreeza committed:
Fix generate invitation test

[Issue #1] https://github.com/iec-uestc/tsemple/issues/1

https://trello.com/c/vjLL1u9r

Fix test 'should generate invitation in limit' so the test would pass

CptBreeza committed:
Change the test for UsersController

[Issue #1] https://github.com/iec-uestc/tsemple/issues/1

https://trello.com/c/vjLL1u9r

Change the test for UsersController that add invitation_code to the
attributes for user so the test would pass

CptBreeza committed:
Improve generate_invitation function

[Issue #1] https://github.com/iec-uestc/tsemple/issues/1

https://trello.com/c/vjLL1u9r

Improve the genereate_invitation function so that it would return the
invitation object instead of nil

CptBreeza committed:
Improve routes and the controller of invitation

[Issue #1] https://github.com/iec-uestc/tsemple/issues/1

https://trello.com/c/vjLL1u9r

Improve the routes and the controller of invitation so that it fits the
RESTful better

CptBreeza committed:
Improve invitation pages with Ajax

[Issue #1] https://github.com/iec-uestc/tsemple/issues/1

https://trello.com/c/vjLL1u9r

Change the invitation pages to Ajax

CptBreeza committed:
Improve details of invitation page

[Issue #1] https://github.com/iec-uestc/tsemple/issues/1

https://trello.com/c/vjLL1u9r

Disable the create button when user's reaching the limit count of
invitations and fix the problem the user who haven't create any
invitation cannot trigger the Ajax.

CptBreeza committed:
Improve locales for invitation

[Issue #1] https://github.com/iec-uestc/tsemple/issues/1

https://trello.com/c/vjLL1u9r

Improve locales for invitation pages and the layout

CptBreeza committed:
Add icons showing the availablity of invitation

[Issue #1] https://github.com/iec-uestc/tsemple/issues/1

https://trello.com/c/vjLL1u9r

Add a lock icon to the invitation which is not available

CptBreeza committed:
Display invitations in the order of id

[Issue #1] https://github.com/iec-uestc/tsemple/issues/1

https://trello.com/c/vjLL1u9r

Display invitations in the order of id.

CptBreeza committed:
Improve the invitation area

[Issue #1] https://github.com/iec-uestc/tsemple/issues/1

https://trello.com/c/vjLL1u9r

Add animation to the invitation area

CptBreeza committed:
Add share icon to the invitation page

[Issue #1] https://github.com/iec-uestc/tsemple/issues/1

https://trello.com/c/vjLL1u9r

Add a share icon and a QRcode area to the invitation row

CptBreeza committed:
Add QRbox div

[Issue #1] https://github.com/iec-uestc/tsemple/issues/1

https://trello.com/c/vjLL1u9r

Add the Qrbox div and display it when the mouse is over the share icon

CptBreeza committed:
Add jQuery.qrcode

[Issue #1] https://github.com/iec-uestc/tsemple/issues/1

https://trello.com/c/vjLL1u9r

Add jQuery.qrcode to generate QRcode for share of invitation code

CptBreeza committed:
Generate QRcode of invitation sharing link

[Issue #1] https://github.com/iec-uestc/tsemple/issues/1

https://trello.com/c/vjLL1u9r

QRcode for invitation sharing link will be generated and displayed when
the mouse is over the share icon

CptBreeza committed:
Add invitation show page

[Issue #1] https://github.com/iec-uestc/tsemple/issues/1

https://trello.com/c/vjLL1u9r

Add a basical show page of invitation code for mobile terminal

CptBreeza committed:
Improve showpage of invitation and fix a bug

[Issue #1] https://github.com/iec-uestc/tsemple/issues/1

https://trello.com/c/vjLL1u9r

Add a Sign up button to allow user to sign up quickly with the given
invitation code.Fix the bug of the Chinese locale.

CptBreeza committed:
Update locales

[Issue #1] https://github.com/iec-uestc/tsemple/issues/1

https://trello.com/c/vjLL1u9r

Update locales for the showpage of invitations

CptBreeza committed:
Remove invitation from Admin

[Issue #1] https://github.com/iec-uestc/tsemple/issues/1

https://trello.com/c/vjLL1u9r

Remove invitation module from Admin

CptBreeza committed:
Add support for admin

[Issue #1] https://github.com/iec-uestc/tsemple/issues/1

https://trello.com/c/vjLL1u9r

Now Administrator doesn't need to turn to Admin page to generate
invitation code.:

CptBreeza committed:
Improve showpage for mobile

[Issue #1] https://github.com/iec-uestc/tsemple/issues/1

https://trello.com/c/vjLL1u9r

Improve the showpage of invitation for mobile so that the lock icon and
share icon will not be showed on mobile

CptBreeza committed:
Improve the route of showpage of invitation

[Issue #1] https://github.com/iec-uestc/tsemple/issues/1

https://trello.com/c/vjLL1u9r

Improve the route of showpage of invitation to be safer than before

CptBreeza committed:
Add index to Invitations code

[Issue #1] https://github.com/iec-uestc/tsemple/issues/1

https://trello.com/c/vjLL1u9r

Add index to Invitations code

CptBreeza committed:
Change the way showing invitations

[Issue #1] https://github.com/iec-uestc/tsemple/issues/1

https://trello.com/c/vjLL1u9r

Now invitations used are placed in the bottom of the showpage

fahchen commented:
@hw676018683 review 一下这个 PR

fahchen commented on app/models/invitation.rb:
```diff
@@ -0,0 +1,17 @@
+  def generate_code
+    begin
+      self.code = SecureRandom.urlsafe_base64
+    end while self.class.where(code: self.code).any?
```
可以试试 `self.class.exists?(code: self.code)`

fahchen commented on db/migrate/20150730083628_add_index_to_invitations_code.rb:
```diff
@@ -0,0 +1,4 @@
+class AddIndexToInvitationsCode < ActiveRecord::Migration
+  def change
+  end
```
这个 migration 是什么意思？

fahchen commented on db/schema.rb:
```diff
@@ -0,0 +1,135 @@
+#
+# It's strongly recommended that you check this file into your version control system.
+
+ActiveRecord::Schema.define(version: 20141127082413) do
```
这里的 version 应该对应的是你最后一个的 migration 的时间戳才对

fahchen commented on app/models/user.rb:
```diff
@@ -78,4 +89,31 @@ def self.find_by_confirmation_token(token)
+    if invitation.any?
+      invitation = invitation.first
+      invitation.available
+    end
```
``` ruby
Invitation.find_by(code: invitation_code).try(:available)
```

可以看下这篇文章 https://ruby-china.org/topics/5348

fahchen commented on app/models/user.rb:
```diff
@@ -78,4 +89,31 @@ def self.find_by_confirmation_token(token)
+
+  def set_invitation
+    invitation = Invitation.where(:code => invitation_code).first
+    invitation.update_attributes(available: false, invitee: self) if !!invitation
```
- [ ] 这里一般很少用 `!!`，除非你只期待 `true` 或 `false`
- [ ] 这里要注意一下，这里的返回值可能会导致你的 callbacks 中断，如果 `invitation.update_attributes(available: false, invitee: self)` 返回 false 时，对 `before_callbacks` 来说是会中断的，`after_callbacks` 不会有问题。当不确定返回值时，最好是给 callbacks 的确定的返回值，true 或 false
  看下这段 http://guides.rubyonrails.org/active_record_callbacks.html#halting-execution

fahchen commented on dump.rdb:
```diff
@@ -0,0 +1 @@
+REDIS0006�ܳC�Z��V
```
删除这个文件

fahchen commented on test/controllers/users/invitations_controller_test.rb:
```diff
@@ -0,0 +1,29 @@
+
+  test "should not generate invitation code out of limit" do
+    session[:return_to] = '/foo'
+    @user.invitation_limit = 0
```
这里应该更新到数据库里面去吧

``` ruby
@user.update invitation_limit: 0
```

fahchen commented on test/controllers/users_controller_test.rb:
```diff
@@ -14,7 +18,7 @@ class UsersControllerTest < ActionController::TestCase
     assert !login?
     assert_difference ["User.count", "ActionMailer::Base.deliveries.size"] do
-      post :create, user: attributes_for(:user)
+      post :create, user: attributes_for(:user).merge(invitation_code: @invitation.code)
```
你可以把这个 `attributes_for(:user).merge(invitation_code: @invitation.code)` 直接扔到 `setup` 里面去

fahchen commented:
<img width="913" alt="screen shot 2015-08-18 at 17 35 33" src="https://cloud.githubusercontent.com/assets/855678/9327040/b417b0ae-45d0-11e5-8ed7-898afdbce2d7.png">
- [ ] 当我没有被使用过的邀请码时，隐藏那块 dom 吧，现在看上去多出来一块
- [ ] 二维码被挡住了，你试试看用 bootstrap 的 tooltip 或 popover 来做
- [ ] 鼠标 hover 邀请码时，二维码会放大撑开整个 dom，这时已经让整个二维码的 panel 发生了抖动，这个是不好的，你要么不放大，要么放大但是不要让整个 panel 抖动放大

fahchen commented:
<img width="1227" alt="screen shot 2015-08-18 at 17 42 04" src="https://cloud.githubusercontent.com/assets/855678/9327151/556ad440-45d1-11e5-96a8-44511f2c7ae5.png">
<img width="1878" alt="screen shot 2015-08-18 at 17 41 55" src="https://cloud.githubusercontent.com/assets/855678/9327154/5877718e-45d1-11e5-9190-213a6d317662.png">
这里的东西布局you wen有问题
- [ ] 布局上整体用 `col-md-8` 居中就差不多了，现在这个太宽了
- [ ] 邀请码用 `.text-center` + `.label` 来做，http://getbootstrap.com/css/#type-alignment ，http://getbootstrap.com/components/#labels
- [ ] 按钮的话，你用 `btn-block` 来做 http://getbootstrap.com/css/#buttons-sizes

CptBreeza committed:
Improve User model

[Issue #1] https://github.com/iec-uestc/tsemple/issues/1

https://trello.com/c/vjLL1u9r

Improve User model

CptBreeza committed:
Improve tests for Invitation and User

[Issue #1] https://github.com/iec-uestc/tsemple/issues/1

https://trello.com/c/vjLL1u9r

Improve tests for Invitation and User

CptBreeza committed:
Remove dump.rdb

CptBreeza committed:
Add index to Invitation code

[Issue #1] https://github.com/iec-uestc/tsemple/issues/1

https://trello.com/c/vjLL1u9r

Add index to Invitation code

CptBreeza committed:
Improve Invitation model

[Issue #1] https://github.com/iec-uestc/tsemple/issues/1

https://trello.com/c/vjLL1u9r

Improve Invitation model

fahchen commented:
改完了 at 我一下

CptBreeza commented:
好的

On Wed, Aug 19, 2015 at 12:39 PM, Phil Chen notifications@github.com
wrote:

> 改完了 at 我一下
> 
> —
> Reply to this email directly or view it on GitHub
> https://github.com/iec-uestc/tsemple/pull/20#issuecomment-132443793.

CptBreeza committed:
Improve the index page of invitation

[Issue #1] https://github.com/iec-uestc/tsemple/issues/1

https://trello.com/c/vjLL1u9r

Improve the index page of invitation

CptBreeza committed:
Improve the show page of invitation

[Issue #1] https://github.com/iec-uestc/tsemple/issues/1

https://trello.com/c/vjLL1u9r

Improve the show page of invitation

CptBreeza committed:
Not zoom the invitation row

[Issue #1] https://github.com/iec-uestc/tsemple/issues/1

https://trello.com/c/vjLL1u9r

Not zoom the invitation row

CptBreeza committed:
Modify z-index of qrbox

[Issue #1] https://github.com/iec-uestc/tsemple/issues/1

https://trello.com/c/vjLL1u9r

Modify the z-index of the qrbox that the qrbox will show above the
navbar

CptBreeza committed:
Click the invitation row to redirect

[Issue #1] https://github.com/iec-uestc/tsemple/issues/1

https://trello.com/c/vjLL1u9r

Now after clicking on the invitation row,a new window will be opened to
show the show page of invitation

CptBreeza committed:
Show the number of invitations left correctly

[Issue #1] https://github.com/iec-uestc/tsemple/issues/1

https://trello.com/c/vjLL1u9r

Fix the bug that the number of invitations left doesn't show correctly

CptBreeza commented:
@fahchen 这块改完了

CptBreeza committed:
Change the position of QRbox when overflowed

[Issue #1] https://github.com/iec-uestc/tsemple/issues/1

https://trello.com/c/vjLL1u9r

Show the QRbox under the invitation row if it is overflowed

fahchen commented on app/views/users/invitations/create.js.erb:
```diff
@@ -0,0 +1,10 @@
+
+<% if !@user.admin? && (@user.invitation_limit - @user.invitations.count == 0) %>
+$('.btn-success').attr('disabled', 'disabled')
+<% end %>
```
这个文件的缩紧

fahchen commented on app/views/users/invitations/index.html.slim:
```diff
@@ -0,0 +1,37 @@
+                  - unless @user.admin?
+                    p.pull-right.invitation_info
+                      = t '.invitations_left', number: @user.invitation_limit - @user.invitations.count
+
```
这个文件写得很乱，`.panel` 嵌套 `.panel` 然后你自己有加了不少类
这个地方重新写一下，全部用 bootstrap 得东西，不要自己定义样式
可以参考下面我稍微修改了一下的样子
<img width="913" alt="screen shot 2015-08-20 at 16 28 17" src="https://cloud.githubusercontent.com/assets/855678/9378989/8e9b2070-4758-11e5-8821-72b62df03540.png">

fahchen replied:
- http://getbootstrap.com/components/#panels-footer
- http://getbootstrap.com/components/#panels-list-group
